pkg/kube/client.go

```diff
@@ -674,48 +944,95 @@ func createPatch(target *resource.Info, current runtime.Object, threeWayMergeFor
-		if err != nil {
-			return fmt.Errorf("failed to create patch: %w", err)
-		}
+	helper := resource.NewHelper(target.Client, target.Mapping).
+		WithFieldValidation(string(fieldValidationDirective)).
+		WithFieldManager(getManagedFieldsManager())
 
-		if patch == nil || string(patch) == "{}" {
-			slog.Debug("no changes detected", "kind", kind, "name", target.Name)
-			// This needs to happen to make sure that Helm has the latest info from the API
-			// Otherwise there will be no labels and other functions that use labels will panic
-			if err := target.Get(); err != nil {
-				return fmt.Errorf("failed to refresh resource information: %w", err)
-			}
-			return nil
-		}
-		// send patch to server
-		slog.Debug("patching resource", "kind", kind, "name", target.Name, "namespace", target.Namespace)
-		obj, err = helper.Patch(target.Namespace, target.Name, patchType, patch, nil)
-		if err != nil {
-			return fmt.Errorf("cannot patch %q with kind %s: %w", target.Name, kind, err)
+	obj, err := helper.Replace(target.Namespace, target.Name, true, target.Object)
+	if err != nil {
+		return fmt.Errorf("failed to replace object: %w", err)
+	}
+
+	if err := target.Refresh(obj, true); err != nil {
+		return fmt.Errorf("failed to refresh object after replace: %w", err)
+	}
+
+	return nil
+
+}
+
+func patchResourceClientSide(original runtime.Object, target *resource.Info, threeWayMergeForUnstructured bool) error {
+
+	patch, patchType, err := createPatch(original, target, threeWayMergeForUnstructured)
+	if err != nil {
+		return fmt.Errorf("failed to create patch: %w", err)
+	}
+
+	kind := target.Mapping.GroupVersionKind.Kind
+	if patch == nil || string(patch) == "{}" {
+		slog.Debug("no changes detected", "kind", kind, "name", target.Name)
+		// This needs to happen to make sure that Helm has the latest info from the API
+		// Otherwise there will be no labels and other functions that use labels will panic
+		if err := target.Get(); err != nil {
+			return fmt.Errorf("failed to refresh resource information: %w", err)
 		}
+		return nil
+	}
+
+	// send patch to server
+	slog.Debug("patching resource", "kind", kind, "name", target.Name, "namespace", target.Namespace)
+	helper := resource.NewHelper(target.Client, target.Mapping).WithFieldManager(getManagedFieldsManager())
+	obj, err := helper.Patch(target.Namespace, target.Name, patchType, patch, nil)
+	if err != nil {
+		return fmt.Errorf("cannot patch %q with kind %s: %w", target.Name, kind, err)
 	}
 
 	target.Refresh(obj, true)
+
 	return nil
 }
 
+// Patch reource using server-side apply
+func patchResourceServerSide(target *resource.Info, dryRun bool, forceConflicts bool, fieldValidationDirective FieldValidationDirective) error {
+	helper := resource.NewHelper(
+		target.Client,
+		target.Mapping).
+		DryRun(dryRun).
+		WithFieldManager(ManagedFieldsManager).
+		WithFieldValidation(string(fieldValidationDirective))
+
```

Comment:
No debug prints in here?

Reply:
I put them in the caller (where more info can be included, and the logging can be separate from the impl):

https://github.com/helm/helm/blob/b2dc411f9d77f3bca969fb4ad955d4a091aa0454/pkg/kube/client.go#L282
https://github.com/helm/helm/blob/b2dc411f9d77f3bca969fb4ad955d4a091aa0454/pkg/kube/client.go#L714